File: services/scoring/app/utils/logger.py
import json
import time
import uuid
from typing import Any

SERVICE_NAME = "scoring"


class StructuredLogger:
    """
    Structured JSON logger that emits to stdout for collection by Docker / k8s log aggregators.
    Every log line includes mandatory PMOS observability fields.
    """

    def __init__(self, layer: str = "service"):
        self.layer = layer

    def _emit(self, level: str, message: str, **extra: Any) -> None:
        record = {
            "level": level,
            "message": message,
            "service": SERVICE_NAME,
            "layer": self.layer,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            **extra,
        }
        print(json.dumps(record), flush=True)

    def info(self, message: str, **extra: Any) -> None:
        self._emit("INFO", message, **extra)

    def warning(self, message: str, **extra: Any) -> None:
        self._emit("WARNING", message, **extra)

    def error(self, message: str, **extra: Any) -> None:
        self._emit("ERROR", message, **extra)

    def debug(self, message: str, **extra: Any) -> None:
        self._emit("DEBUG", message, **extra)

    def critical(self, message: str, **extra: Any) -> None:
        self._emit("CRITICAL", message, **extra)


def new_span_id() -> str:
    return str(uuid.uuid4())


# Default module-level logger
logger = StructuredLogger(layer="service")

File: services/gateway/app/utils/logger.ts
interface LogRecord {
  level: string;
  message: string;
  service: 'gateway';
  timestamp: string;
  trace_id?: string;
  span_id?: string;
  layer?: string;
  duration_ms?: number;
  [key: string]: unknown;
}

type LogLevel = 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';

const LOG_LEVEL_PRIORITY: Record<LogLevel, number> = {
  DEBUG: 0,
  INFO: 1,
  WARN: 2,
  ERROR: 3,
};

function getCurrentLogLevel(): LogLevel {
  const level = process.env['LOG_LEVEL'] as LogLevel | undefined;
  return level && LOG_LEVEL_PRIORITY[level] !== undefined ? level : 'INFO';
}

function shouldLog(level: LogLevel): boolean {
  return LOG_LEVEL_PRIORITY[level] >= LOG_LEVEL_PRIORITY[getCurrentLogLevel()];
}

function emit(level: LogLevel, message: string, extra: Partial<LogRecord> = {}): void {
  if (!shouldLog(level)) return;
  const record: LogRecord = {
    level,
    message,
    service: 'gateway',
    timestamp: new Date().toISOString(),
    ...extra,
  };
  // stdout only — collected by Docker/k8s log aggregator
  console.log(JSON.stringify(record));
}

export const logger = {
  debug(message: string, extra?: Partial<LogRecord>): void {
    emit('DEBUG', message, extra);
  },
  info(message: string, extra?: Partial<LogRecord>): void {
    emit('INFO', message, extra);
  },
  warn(message: string, extra?: Partial<LogRecord>): void {
    emit('WARN', message, extra);
  },
  error(message: string, extra?: Partial<LogRecord>): void {
    emit('ERROR', message, extra);
  },
};

export type { LogRecord };
